content/blog/posts/c-stl-algorithms-series-part-2.md

```diff
@@ -0,0 +1,69 @@
+
+My initial intuition was to try and use `minmax_element` or some combination of `min` and `max` from `<algorithm>`, but that proved to be non-starter. Those algorithms are implemented in terms of the `<` operator which doesn’t fit particularly well with `Vector3` and the overload still requires a binary comparison function. Back to the drawing board...
+
+Thinking about the problem again, we’re essentially going from many to one, and there’s an algorithm that’s perfect for that - `accumulate`. `accumulate` turns out to be incredibly versatile. In this example we provide the range to operate on, the initial value and a _reduction_ function. The lambda/function for the reduction is the interesting part. The first parameter is the accumulator value (`acc` by convention), the second parameter is the current element in the range we’re operating on, and the return value is the next accumulator.
+
+```c++
+AZStd::vector<AZ::Aabb> aabbs { /* a bunch of Aabbs */ };
+const auto aabbMinMax = AZStd::accumulate(
+    aabbs.cbegin(), aabbs.cend(), AZ::Aabb::CreateNull(),
+    [](AZ::Aabb acc, const AZ::Aabb& aabb)
+    {
+        return AZ::Aabb{
+            acc.GetMin().GetMin(aabb.GetMin()),
+            acc.GetMax().GetMax(aabb.GetMax())};
+    });
+```
+
+## Deliberation
+
+It’s debatable if the algorithm version is more succinct than the first, but it does have some advantages. First, the return value can now be `const`, as everything is calculated in a single statement. Eliminating mutable state is well understood to reduce the likelihood of bugs creeping into code. The lambda is a pure function and due to having no capture (highly recommended) we can quickly determine it has no side effects. The other advantage is the name, we can glance at the code and quickly get an intuition for what operation is taking place (even if we have to check the types and lambda to understand all the details).
+
+Applications of `accumulate` and `transform` (look out for this in an upcoming article) crop up all over the place and making use of them can help simplify and improve our code. (Note: `accumulate` is very closely related to `reduce` - they are largely the same but do have some differences when it comes to ordering and execution policies. This isn’t important for understanding what they actually do).
+
+## Further Reading
+
+If you’re getting more interested in algorithms the next must-watch talk is
+
+> [Algorithm Intuition](https://youtu.be/48gV1SNm3WA) by Conor Hoekstra
+
+The talk does a fantastic job of introducing a whole host of algorithms and how to think about them.
+
+## To be continued...
+
+Next time we’ll look at another algorithm staple - `transform`.
+
+Author: Tom Hulton-Harrop
+
+_Disclaimer: The views expressed here are those of the individual author and do not represent those of Game Engine and Developer Services or Amazon._
```

Comment:
Same comment as the PR for Part 1 regarding disclaimers.

Reply:
I will add the new one. The new one will be 
Disclaimer: The views expressed here are those of the individual author and do not represent those of the Open 3D Foundation, Open 3D Engine or individual's respective company.